airflow/utils/file.py

```diff
@@ -215,21 +215,21 @@ def _find_path_from_directory(
 
         ignore_file_path = Path(root) / ignore_file_name
         if ignore_file_path.is_file():
-            with open(ignore_file_path) as ifile:
-                lines_no_comments = [re2.sub(r"\s*#.*", "", line) for line in ifile.read().split("\n")]
-                # append new patterns and filter out "None" objects, which are invalid patterns
-                patterns += [
-                    p
-                    for p in [
-                        ignore_rule_type.compile(line, Path(base_dir_path), ignore_file_path)
-                        for line in lines_no_comments
-                        if line
-                    ]
-                    if p is not None
+            lines = ignore_file_path.read_text().splitlines()
```

Comment:
I would keep the `with open` style and rewrite the reading part to

```python
[re2.sub(r"\s*#.*", "", line) for line in ifile]
```

This would allow the input to be read line by line instead of creating a temporary string before splitting.

Reply:
How about

```python
        ignore_file_path = Path(root) / ignore_file_name
        if ignore_file_path.is_file():
            uniq_patterns = {}
            with ignore_file_path.open() as f:
                for line in f:
                    line = line.strip()
                    if line and not line.startswith("#"):
                        pattern = ignore_rule_type.compile(line, Path(base_dir_path), ignore_file_path)
                        if pattern:
                            uniq_patterns[pattern] = None
            # evaluation order of patterns is important with negation
            # so that later patterns can override earlier patterns
            patterns = list(uniq_patterns)
```

This probably describes the steps even better, and shows that the `ignore_rule_type.compile` is called before the patterns are filtered via `uniq` (dict).

Reply:
Yeah I wouldn’t object.